javalib/src/main/scala/java/util/Map.scala

```diff
@@ -15,10 +21,159 @@ trait Map[K, V] {
   def entrySet(): Set[Map.Entry[K, V]]
   def equals(o: Any): Boolean
   def hashCode(): Int
+
+  // Java 8 default methods
+
+  def getOrDefault(key: Any, default: V): V = {
+    val v = this.get(key)
+    if ((v == null.asInstanceOf[V]) && this.containsKey(key))
+      null.asInstanceOf[V]
```

Comment:
I think this should be 
```scala
if ((v ne null ) || this.containsKey(key)) v.asInstanceOf[V] else default
```

Reply:
or just:
```
if(this.containsKey(key)) this.get(key) else default
```
?

Reply:
Will scala-native has the `@JavaDefaultMethod` annotation?

Reply:
refs: https://github.com/scala-js/scala-js/blob/master/javalib/src/main/scala/java/util/Map.scala#L38

Reply:
@hepin1989 
Thank you for your review and the suggestions.

I encountered @JavaDefaultMethod recently whilst porting another file from Scala.js.  I am only a grasshopper
but I saw nothing essential in Scala Native for the annotation. It would be make porting easier, especially porting
 back to Scala.js.  I think in the file I looked at, I had to delete two lines.  That could be done by an awk script
in the Sjs->SN direction but, I think, would be harder in the reverse direction.  Having no need and many other
requests/demands/fires, I did not take the time to implement.  Now, when the Big Dogs review that, I may learn
of its necessity,.

Reply:
@hepin1989  Your second suggestion is, with slightly different formatting, what is done in Scala.js.

That PR was done independently by a better engineer that I.  The PR was created about three days
before the Scala.js PR.  I did not know that the latter was in-flight. The original work on this PR was 
done over a year or so ago.

For consistency between the cousins, I can change this PR to use the Scala.js code.

At the time I wrote this, I was concerned about execution time efficiency, as I almost always am.
If we are leading people to use a method, it should be reasonably efficient or else subclasses should
override it.

In the Java 8 design, Map is extended by various sub-classes and the sub-classes have different
performance profiles for get() and contains(). Although I am currently implementing TreeMap for SN,
I believe the most commonly used Maps are various flavors of HashMap.  Reasonable implementations
of HashMap have an Order of one (one access) O(1) cost, meaning they are fast (yes, I know about
lists in hash buckets & degenerate cases).  If I remember my
ancient Analysis of Algorithms course correctly, contains() has a worst case performance based on
the number of elements N. that is Order of N, or O(N). That is on "key not found". I believe that the
expected performance where the key is likely to be in the table is O(N/2).

Now, for reasonably large Maps (say 30 or 300) TreeMap has contains of O(logN) but get() is also O(logN).
Not bad for larger numbers, but not O(1), (ignoring or with limited hash conflicts).

By my reading and intent, the code of this PR is optimized for the "Key found" case.  It does the cheap
get() method first and only falls back to the expensive contains() method when the value is either not found 
or an actual null, attached to a key.

This find tuning of expected execution behavior is somewhat of a tempest in a teapot because
many implementations will override the default method.

I believe the code as written is correct. I can modify either it or the Scala.js code as recommended
by senior reviewers. Sometimes there are higher values than examination question optimizations.

Reply:
Yeah，I think porting from scala.ja with some automatic tools would be much easier too。then we can first send pr to scala.js and sync it back here。

Reply:
But yes, I would like to pass the examination question ;-)

Also insert usual caveats about optimizations done in anticipation but without benchmarks
and about expected workloads & sizes without backing data.

Any corect variant of the method is likely to save end users precious development time and 
is better than what went before.

Reply:
Given what I know now, I would probably not make point edits but just port the Scala.js variant.
I might investigate and take the time to implement @JavaDefaultMethod annotation as a no-op,
so that future porting might require fewer changes.

Please advise.

Reply:
> But yes, I would like to pass the examination question ;-)
> 
> Also insert usual caveats about optimizations done in anticipation but without benchmarks
> and about expected workloads & sizes without backing data.
> 
> Any corect variant of the method is likely to save end users precious development time and 
> is better than what went before.

When will the `v` which already associated with the key be return in the above implementation ？

Reply:
@hepin1989  


I think right now we, the Scala Native community, need to learn a bit more about what can be done to
ease porting, both back & forth.   I believe that automatic tools are in the mind of a number of people, but,
speaking for myself, I must walk through a process a couple of time (well, a couple of tens of times)
before I can automate it.  Each time I try to reduce grunt work, both for me and others.  But I am
wary of "Garbage in; Garbage out!".

Thinking aloud, perhaps one or more scripts to comment out annotations used in Scala.js but
not in Scala Native would be a better way to go than implementing no-op annotations.
I know that leaving commented out code lying around is _highly_ discouraged but Scala Native
does have some commented out test code.  Maybe I will run that up the flagpole.

Reply:
> Given what I know now, I would probably not make point edits but just port the Scala.js variant.
> I might investigate and take the time to implement @JavaDefaultMethod annotation as a no-op,
> so that future porting might require fewer changes.
> 
> Please advise.

+100 for this。I more like to see the missing javalib contributions first get into scala.js，and then can be easily been ported to here , maybe sync automatically？

Reply:
"sync automatically" is a good goal, but I think it years out.  In other projects I see scala-steward saving people
a lot of grunt work keeping up with libraryDependency changes.

There has been a lot of work in Scala.js, especially over the past two weeks, that I look at and say "Oops, it would be great if I could port that to Scala Native", just run out of time.

Reply:
> When will the v which already associated with the key be return in the above implementation ？
You are right, it will not! Oops!  Thank you for catching that.

While the re-port discussion is up in the air, I will fix getOrDefault(), if nothing else
that to reduce my embarrassment.

I will also check the unit-test to see why that blunder was not caught.

Lastly, I will check the Scala.js Map.scala to see if all the default methods are implemented.

Looks like I have my Saturday's work cut out for me.

Reply:
This discussion is really long but some Maps allow null keys and values. The first looked like a check for that.